File: app/src/main/java/org/torproject/android/ui/AppManagerActivity.java
/* Copyright (c) 2009, Nathan Freitas, Orbot / The Guardian Project - http://openideals.com/guardian */
/* See LICENSE for licensing information */

package org.torproject.android.ui;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Comparator;
import java.util.Iterator;
import java.util.List;
import java.util.StringTokenizer;

import org.torproject.android.service.OrbotConstants;
import org.torproject.android.R;
import org.torproject.android.service.util.TorServiceUtils;
import org.torproject.android.service.vpn.TorifiedApp;

import android.app.ProgressDialog;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.SharedPreferences.Editor;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.os.AsyncTask;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.View.OnClickListener;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.CheckBox;
import android.widget.CompoundButton;
import android.widget.CompoundButton.OnCheckedChangeListener;
import android.widget.GridView;
import android.widget.ImageView;
import android.widget.ListAdapter;
import android.widget.TextView;

public class AppManagerActivity extends AppCompatActivity implements OnClickListener, OrbotConstants {

    private GridView listApps;
    private ListAdapter adapterApps;
    private final static String TAG = "Orbot";
    PackageManager pMgr = null;

    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        pMgr = getPackageManager();

        this.setContentView(R.layout.layout_apps);
        setTitle(R.string.apps_mode);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);

    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == android.R.id.home)
        {
            finish();
            return true;
        }

        return super.onOptionsItemSelected(item);
    }

    @Override
    protected void onResume() {
        super.onResume();
        listApps = (GridView) findViewById(R.id.applistview);
        mPrefs = TorServiceUtils.getSharedPrefs(getApplicationContext());


        new AsyncTask<Void, Void, Void>() {
            private ProgressDialog dialog;

            protected void onPreExecute() {
                // Pre Code
                dialog = new ProgressDialog(AppManagerActivity.this, android.support.v4.app.DialogFragment.STYLE_NO_TITLE);
                dialog.show();
            }
            protected Void doInBackground(Void... unused) {
                loadApps(mPrefs);
                return null;
            }
            protected void onPostExecute(Void unused) {
                listApps.setAdapter(adapterApps);
                dialog.cancel();
            }
        }.execute();


    }

    SharedPreferences mPrefs = null;
    ArrayList<TorifiedApp> mApps = null;

    private void loadApps (SharedPreferences prefs)
    {

        mApps = getApps(getApplicationContext(), prefs);
        Collections.sort(mApps,new Comparator<TorifiedApp>() {
            public int compare(TorifiedApp o1, TorifiedApp o2) {
                if (o1.isTorified() == o2.isTorified()) return o1.getName().compareTo(o2.getName());
                if (o1.isTorified()) return -1;
                return 1;
            }
        });

        final LayoutInflater inflater = getLayoutInflater();

        adapterApps = new ArrayAdapter<TorifiedApp>(this, R.layout.layout_apps_item, R.id.itemtext,mApps) {

            public View getView(int position, View convertView, ViewGroup parent) {

                ListEntry entry = null;

                if (convertView == null)
                    convertView = inflater.inflate(R.layout.layout_apps_item, parent, false);
                else
                    entry = (ListEntry) convertView.getTag();

                if (entry == null) {
                    // Inflate a new view
                    entry = new ListEntry();
                    entry.icon = (ImageView) convertView.findViewById(R.id.itemicon);
                    entry.box = (CheckBox) convertView.findViewById(R.id.itemcheck);
                    entry.text = (TextView) convertView.findViewById(R.id.itemtext);
                    convertView.setTag(entry);
                }

                final TorifiedApp app = mApps.get(position);

                if (entry.icon != null) {

                    try {
                        entry.icon.setImageDrawable(pMgr.getApplicationIcon(app.getPackageName()));
                        entry.icon.setOnClickListener(AppManagerActivity.this);
                        entry.icon.setTag(entry.box);
                    }
                    catch (Exception e)
                    {
                        e.printStackTrace();
                    }
                }

                if (entry.text != null) {
                    entry.text.setText(app.getName());
                    entry.text.setOnClickListener(AppManagerActivity.this);
                    entry.text.setTag(entry.box);
                }

                if (entry.box != null) {
                    entry.box.setOnClickListener(AppManagerActivity.this);
                    entry.box.setChecked(app.isTorified());
                    entry.box.setTag(app);
                }

                return convertView;
            }
        };


    }

    private static class ListEntry {
        private CheckBox box;
        private TextView text;
        private ImageView icon;
    }

    /* (non-Javadoc)
     * @see android.app.Activity#onStop()
     */
    @Override
    protected void onStop() {
        super.onStop();

    }


    public ArrayList<TorifiedApp> getApps (Context context, SharedPreferences prefs)
    {

        String tordAppString = prefs.getString(PREFS_KEY_TORIFIED, "");
        String[] tordApps;

        StringTokenizer st = new StringTokenizer(tordAppString,"|");
        tordApps = new String[st.countTokens()];
        int tordIdx = 0;
        while (st.hasMoreTokens())
        {
            tordApps[tordIdx++] = st.nextToken();
        }

        List<ApplicationInfo> lAppInfo = pMgr.getInstalledApplications(0);

        Iterator<ApplicationInfo> itAppInfo = lAppInfo.iterator();

        ArrayList<TorifiedApp> apps = new ArrayList<TorifiedApp>();

        ApplicationInfo aInfo = null;

        TorifiedApp app = null;

        while (itAppInfo.hasNext())
        {
            aInfo = itAppInfo.next();

            app = new TorifiedApp();

            try {
                PackageInfo pInfo = pMgr.getPackageInfo(aInfo.packageName, PackageManager.GET_PERMISSIONS);

                if (pInfo != null && pInfo.requestedPermissions != null)
                {
                    for (String permInfo:pInfo.requestedPermissions)
                    {
                        if (permInfo.equals("android.permission.INTERNET"))
                        {
                            app.setUsesInternet(true);

                        }
                    }

                }


            } catch (Exception e) {
                // TODO Auto-generated catch block
                e.printStackTrace();
            }

            /**
             if ((aInfo.flags & ApplicationInfo.FLAG_SYSTEM) == 1)
             {
             //System app
             app.setUsesInternet(true);
             }**/


            try
            {
                app.setName(pMgr.getApplicationLabel(aInfo).toString());
            }
            catch (Exception e)
            {
                // no name
                continue; //we only show apps with names
            }


            if (!app.usesInternet())
                continue;
            else
            {
                apps.add(app);
            }

            app.setEnabled(aInfo.enabled);
            app.setUid(aInfo.uid);
            app.setUsername(pMgr.getNameForUid(app.getUid()));
            app.setProcname(aInfo.processName);
            app.setPackageName(aInfo.packageName);


            // check if this application is allowed
            if (Arrays.binarySearch(tordApps, app.getUsername()) >= 0) {
                app.setTorified(true);
            }
            else
            {
                app.setTorified(false);
            }

        }

    //    Collections.sort(apps);

        return apps;
    }


    public void saveAppSettings (Context context)
    {

        StringBuilder tordApps = new StringBuilder();
        Intent response = new Intent();

        for (TorifiedApp tApp:mApps)
        {
            if (tApp.isTorified())
            {
                tordApps.append(tApp.getUsername());
                tordApps.append("|");
                response.putExtra(tApp.getUsername(),true);
            }
        }

        Editor edit = mPrefs.edit();
        edit.putString(PREFS_KEY_TORIFIED, tordApps.toString());
        edit.commit();

        setResult(RESULT_OK,response);
    }


    /**
     * Called an application is check/unchecked
     */
    /**
    public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
        final TorifiedApp app = (TorifiedApp) buttonView.getTag();
        if (app != null) {
            app.setTorified(isChecked);
        }

        saveAppSettings(this);

    }**/




    public void onClick(View v) {

        CheckBox cbox = null;

        if (v instanceof CheckBox)
            cbox = (CheckBox)v;
        else if (v.getTag() instanceof CheckBox)
            cbox = (CheckBox)v.getTag();

        if (cbox != null) {
            final TorifiedApp app = (TorifiedApp) cbox.getTag();
            if (app != null) {
                app.setTorified(!app.isTorified());
                cbox.setChecked(app.isTorified());
            }

            saveAppSettings(this);
        }
    }




}

File: app/src/main/java/org/torproject/android/ui/AppManagerActivity.kt
package org.torproject.android.ui

import android.Manifest
import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.content.pm.ApplicationInfo
import android.content.pm.PackageManager
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.View.OnFocusChangeListener
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.CheckBox
import android.widget.GridView
import android.widget.ImageView
import android.widget.ListAdapter
import android.widget.ProgressBar
import android.widget.TextView
import androidx.appcompat.widget.Toolbar

import androidx.core.content.ContextCompat
import androidx.core.content.edit

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.FlowPreview
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.debounce
import kotlinx.coroutines.flow.distinctUntilChanged
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

import org.torproject.android.BuildConfig
import org.torproject.android.R
import org.torproject.android.service.OrbotConstants
import org.torproject.android.service.util.Prefs
import org.torproject.android.service.vpn.TorifiedApp
import org.torproject.android.service.vpn.TorifiedAppWrapper
import org.torproject.android.ui.core.BaseActivity

import java.util.Arrays
import java.util.StringTokenizer

@OptIn(FlowPreview::class)
class AppManagerActivity : BaseActivity(), View.OnClickListener {

    private var pMgr: PackageManager? = null
    private var mPrefs: SharedPreferences? = null
    private var listAppsAll: GridView? = null
    private var adapterAppsAll: ListAdapter? = null
    private var progressBar: ProgressBar? = null
    private var alSuggested: List<String>? = null
    private var searchBar: TextView? = null
    private var filteredList: MutableList<TorifiedAppWrapper> = ArrayList()
    private val searchQuery = MutableStateFlow("")
    private var retainedCheckedPackages: Set<String> = emptySet()

    private val job = Job()
    private val scope = CoroutineScope(Dispatchers.Main + job)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        pMgr = packageManager
        this.setContentView(R.layout.activity_app_manager)


        var toolbar = findViewById<Toolbar>(R.id.toolbar);
        setSupportActionBar(toolbar);
        var sab = getSupportActionBar();
        if (sab != null) sab.setDisplayHomeAsUpEnabled(true);
        
        listAppsAll = findViewById(R.id.applistview)
        progressBar = findViewById(R.id.progressBar)
        searchBar = findViewById(R.id.searchBar)

        retainedCheckedPackages = savedInstanceState?.getStringArray("checked_packages")?.toSet() ?: emptySet()
        val restoredQuery = savedInstanceState?.getString("search_query").orEmpty()
        if (restoredQuery.isNotEmpty()) {
            searchBar?.text = restoredQuery
            searchQuery.value = restoredQuery
        }

        searchQuery
            .debounce(250)
            .distinctUntilChanged()
            .onEach { filterApps(it) }
            .launchIn(scope)

        searchBar?.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
                searchQuery.value = s?.toString().orEmpty()
            }
            override fun afterTextChanged(s: Editable?) {}
        })

        // Need a better way to manage this list
        alSuggested = OrbotConstants.VPN_SUGGESTED_APPS
    }

    override fun onResume() {
        super.onResume()
        mPrefs = Prefs.getSharedPrefs(applicationContext)
        reloadApps()
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putString("search_query", searchQuery.value)
        val checkedPackages = (allApps.orEmpty() + suggestedApps.orEmpty())
            .filter { it.isTorified }
            .map { it.packageName }
            .toTypedArray()
        outState.putStringArray("checked_packages", checkedPackages)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        super.onCreateOptionsMenu(menu)
        val inflater = menuInflater
        inflater.inflate(R.menu.app_main, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == R.id.menu_save_apps) {
            saveAppSettings()
            finish()
        } else if (item.itemId == android.R.id.home) {
            finish()
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    private fun reloadApps() {

        scope.launch {
            progressBar?.visibility = View.VISIBLE
            withContext(Dispatchers.IO) {
                loadApps()
            }
            listAppsAll?.adapter = adapterAppsAll
            progressBar?.visibility = View.GONE

            filterApps(searchQuery.value)
        }
    }

    private fun filterApps(query: String?) {
        scope.launch(Dispatchers.Default) {
            val lower = query?.lowercase()?.trim().orEmpty()
            val results = if (lower.isEmpty()) {
                allUnfilteredUiItems
            } else {
                allUnfilteredUiItems.filter { it.app?.name?.lowercase()?.contains(lower) == true }
            }

            withContext(Dispatchers.Main) {
                filteredList.clear()
                filteredList.addAll(results)
                (adapterAppsAll as? ArrayAdapter<*>)?.notifyDataSetChanged()
            }
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        job.cancel()
    }

    private var allApps: List<TorifiedApp>? = null
    private var suggestedApps: List<TorifiedApp>? = null
    // contains apps, but also other things like TextViews for suggested apps
    var allUnfilteredUiItems: MutableList<TorifiedAppWrapper> = ArrayList()

    private fun loadApps() {
        if (allApps == null) allApps = getApps(this@AppManagerActivity, mPrefs, null, alSuggested)
        TorifiedApp.sortAppsForTorifiedAndAbc(allApps)
        if (suggestedApps == null) suggestedApps =
            getApps(this@AppManagerActivity, mPrefs, alSuggested, null)
        val inflater = layoutInflater
        if (allUnfilteredUiItems.isEmpty()) {
            // only show suggested apps, text, etc and other apps header if there are any suggested apps installed...
            if (!suggestedApps.isNullOrEmpty()) {
                val headerSuggested = TorifiedAppWrapper()
                headerSuggested.header = getString(R.string.apps_suggested_title)
                allUnfilteredUiItems.add(headerSuggested)
                val subheaderSuggested = TorifiedAppWrapper()
                subheaderSuggested.subheader = getString(R.string.app_suggested_subtitle)
                allUnfilteredUiItems.add(subheaderSuggested)

                allUnfilteredUiItems.addAll(suggestedApps?.map { TorifiedAppWrapper(app = it) } ?: emptyList())

                val headerAllApps = TorifiedAppWrapper()
                headerAllApps.header = getString(R.string.apps_other_apps)
                allUnfilteredUiItems.add(headerAllApps)
            }

            allUnfilteredUiItems.addAll(allApps?.map { TorifiedAppWrapper(app = it) } ?: emptyList())
        }

        adapterAppsAll = object : ArrayAdapter<TorifiedAppWrapper?>(
            this,
            R.layout.layout_apps_item,
            R.id.itemtext,
            filteredList
        ) {
            override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
                var cv = convertView
                var entry: ListEntry? = null

                if (cv == null) {
                    cv = inflater.inflate(R.layout.layout_apps_item, parent, false)
                }
                else {
                    entry = cv.tag as ListEntry
                }

                if (entry == null) {
                    // Inflate a new view
                    entry = ListEntry()
                    entry.container = cv?.findViewById(R.id.appContainer)
                    entry.icon = cv?.findViewById(R.id.itemicon)
                    entry.box = cv?.findViewById(R.id.itemcheck)
                    entry.text = cv?.findViewById(R.id.itemtext)
                    entry.header = cv?.findViewById(R.id.tvHeader)
                    entry.subheader = cv?.findViewById(R.id.tvSubheader)
                    cv?.tag = entry
                }

                val taw = filteredList[position]

                if (taw.header != null) {
                    entry.header?.text = taw.header
                    entry.header?.visibility = View.VISIBLE
                    entry.subheader?.visibility = View.GONE
                    entry.container?.visibility = View.GONE
                }
                else if (taw.subheader != null) {
                    entry.subheader?.visibility = View.VISIBLE
                    entry.subheader?.text = taw.subheader
                    entry.container?.visibility = View.GONE
                    entry.header?.visibility = View.GONE
                }
                else {
                    val app = taw.app
                    entry.header?.visibility = View.GONE
                    entry.subheader?.visibility = View.GONE
                    entry.container?.visibility = View.VISIBLE

                    val packageName = app?.packageName
                    if (entry.icon != null && packageName != null) {
                        try {
                            entry.icon?.setImageDrawable(pMgr?.getApplicationIcon(packageName))
                            entry.icon?.tag = entry.box
                            entry.icon?.setOnClickListener(this@AppManagerActivity)
                        }
                        catch (e: Exception) {
                            e.printStackTrace()
                        }
                    }

                    entry.text?.text = app?.name
                    entry.text?.tag = entry.box
                    entry.text?.setOnClickListener(this@AppManagerActivity)

                    entry.box?.isChecked = app?.isTorified ?: false
                    entry.box?.tag = app
                    entry.box?.setOnClickListener(this@AppManagerActivity)
                }

                cv?.onFocusChangeListener =
                    OnFocusChangeListener { v: View, hasFocus: Boolean ->
                        if (hasFocus) v.setBackgroundColor(
                            ContextCompat.getColor(
                                context, R.color.dark_purple
                            )
                        ) else {
                            v.setBackgroundColor(
                                ContextCompat.getColor(
                                    context,
                                    android.R.color.transparent
                                )
                            )
                        }
                    }

                return cv ?: View(context)
            }
        }

        filteredList.clear()
        filteredList.addAll(allUnfilteredUiItems)
    }

    private fun saveAppSettings() {
        val allApps = allApps ?: return
        val suggestedApps = suggestedApps ?: return

        val tordApps = StringBuilder()
        val response = Intent()

        for (tApp in allApps) {
            if (tApp.isTorified) {
                tordApps.append(tApp.packageName)
                tordApps.append("|")
                response.putExtra(tApp.packageName, true)
            }
        }

        for (tApp in suggestedApps) {
            if (tApp.isTorified) {
                tordApps.append(tApp.packageName)
                tordApps.append("|")
                response.putExtra(tApp.packageName, true)
            }
        }

        mPrefs?.edit {
            putString(OrbotConstants.PREFS_KEY_TORIFIED, tordApps.toString())
        }

        setResult(RESULT_OK, response)
    }

    override fun onClick(v: View) {
        // todo make whatever is clicked (box, icon, text) set selected=true on the TextView
        // to make the text scroll, for now only selecting the text itself makes it scroll
        var cbox: CheckBox? = null
        if (v is CheckBox) cbox = v
        else if (v.tag is CheckBox) cbox = v.tag as CheckBox
        else if (v.tag is ListEntry) cbox = (v.tag as ListEntry).box
        if (cbox != null) {
            val app = cbox.tag as TorifiedApp
            app.isTorified = !app.isTorified
            cbox.isChecked = app.isTorified
        }
    }

    private class ListEntry {
        var box: CheckBox? = null
        var text: TextView? = null // app name
        var icon: ImageView? = null
        var container: View? = null
        var header: TextView? = null
        var subheader: TextView? = null
    }

    companion object {
        /**
         * @return true if the app is "enabled", not Orbot, and not in
         * [.BYPASS_VPN_PACKAGES]
         */
        private fun includeAppInUi(applicationInfo: ApplicationInfo): Boolean {
            return applicationInfo.enabled &&
                   applicationInfo.packageName != BuildConfig.APPLICATION_ID &&
                   !OrbotConstants.BYPASS_VPN_PACKAGES.contains(applicationInfo.packageName)
        }

        fun getApps(
            context: Context,
            prefs: SharedPreferences?,
            filterInclude: List<String>?,
            filterRemove: List<String>?
        ): ArrayList<TorifiedApp> {
            val pMgr = context.packageManager
            val tordAppString = prefs?.getString(OrbotConstants.PREFS_KEY_TORIFIED, "")
            val tordApps: Array<String?>
            val st = StringTokenizer(tordAppString, "|")
            tordApps = arrayOfNulls(st.countTokens())
            var tordIdx = 0
            while (st.hasMoreTokens()) {
                tordApps[tordIdx++] = st.nextToken()
            }
            Arrays.sort(tordApps)
            val lAppInfo = pMgr.getInstalledApplications(0)
            val itAppInfo: Iterator<ApplicationInfo> = lAppInfo.iterator()
            val apps = ArrayList<TorifiedApp>()
            while (itAppInfo.hasNext()) {
                val aInfo = itAppInfo.next()
                if (!includeAppInUi(aInfo)) continue
                if (filterInclude != null) {
                    var wasFound = false
                    for (filterId in filterInclude) if (filterId == aInfo.packageName) {
                        wasFound = true
                        break
                    }
                    if (!wasFound) continue
                }
                if (filterRemove != null) {
                    var wasFound = false
                    for (filterId in filterRemove) if (filterId == aInfo.packageName) {
                        wasFound = true
                        break
                    }
                    if (wasFound) continue
                }
                val app = TorifiedApp()
                try {
                    val pInfo = pMgr.getPackageInfo(aInfo.packageName, PackageManager.GET_PERMISSIONS)

                    for (permInfo in pInfo.requestedPermissions ?: emptyArray()) {
                        if (permInfo == Manifest.permission.INTERNET) {
                            app.usesInternet = true
                        }
                    }
                } catch (e: Exception) {
                    e.printStackTrace()
                }

                try {
                    app.name = pMgr.getApplicationLabel(aInfo).toString()
                } catch (e: Exception) {
                    // No name, we only show apps with names
                    continue
                }

                if (!app.usesInternet) continue else {
                    apps.add(app)
                }

                app.isEnabled = aInfo.enabled
                app.uid = aInfo.uid
                app.username = pMgr.getNameForUid(app.uid)
                app.procname = aInfo.processName
                app.packageName = aInfo.packageName

                // Check if this application is allowed
                app.isTorified = Arrays.binarySearch(tordApps, app.packageName) >= 0

                // Preserve rotation-checked state
                app.isTorified = app.isTorified || (context as? AppManagerActivity)?.retainedCheckedPackages?.contains(app.packageName) == true
            }
            apps.sort()

            return apps
        }
    }
}
